src/qutip_qip/vqa.py

```diff
@@ -0,0 +1,841 @@
+        return format(outcome_index, f"0{n_qubits}b")
+
+    def evaluate_parameters(self, angles):
+        """
+        Evaluate a cost for the circuit, based on the VQA cost
+        method defined.
+
+        Parameters
+        ----------
+        angles: list of float
+            A list of dimension (n,) for n free parameters in the circuit
+
+        Returns
+        -------
+        cost: float
+        """
+        final_state = self.get_final_state(angles)
+        if self.cost_method == "BITSTRING":
+            if self.cost_func is None:
+                raise ValueError("self.cost_func not specified")
+            else:
+                return self.cost_func(
+                    self._sample_bitstring_from_state(final_state)
+                )
+        elif self.cost_method == "STATE":
+            if self.cost_func is None:
+                raise ValueError("self.cost_func not specified")
+            else:
+                return self.cost_func(final_state)
+        elif self.cost_method == "OBSERVABLE":
+            if self.cost_observable is None:
+                raise ValueError("self.cost_observable not specified")
+            else:
+                return expect(self.cost_observable, final_state)
+        else:
+            raise ValueError(f"Unrecognised cost method: {self.cost_method}")
+
+    def optimize_parameters(
+        self,
+        initial="random",
+        method="COBYLA",
+        use_jac=False,
+        do_nothing=False,
+        layer_by_layer=False,
+        bounds=None,
+        constraints=(),
+    ):
+        """
+        Run VQA optimization loop
+
+        Parameters
+        ----------
+        initial: str or list of floats, optional
+            Initialization method for the free parameters.
+            If a list of floats of dimensions (n,) for n free parameters
+            in the circuit is given, then these are taken to be the initial
+            conditions for the optimizer. Otherwise if a string is given:
+            1. "random" will randomize initial free parameters between 0 and 1,
+            2. "ones" will set each initial free parameter to a value of 1.
+            Defaults to "random"
+        method: str or callable, optional
+            Method to give to ``scipy.optimize.minimize``
+        use_jac: bool, optional
+            Whether to compute the jacobian or not. If computed, it will be
+            passed to the optimizer chosen by ``method``, regardless of if
+            the method is gradient-based or not.
+            Note that derivatives of unitaries generated by
+            ``ParameterizedHamiltonian`` are calculated with the
+            Frechet derivative of the exponential function,
+            using ``scipy.linalg.expm_frechet``.
+        do_nothing: bool, optional
+            For debugging and benchmarking purposes, perform no operation
+            during optimization, and simply retain the initial free parameters.
```

Comment:
This is a bit confusing, I understand that for debugging you might just want to evaluate the parameters. However, having this in the optimization method seems contradictory to the name of the method.
Is this necessary? Why not just call the evaluation function in debugging?

Reply:
Oops, I think I meant to remove this! Has been removed now.